### DISCUSSION TITLE
How to specify the destructor function during the extension of MicroPython using C

### DISCUSSION THREAD

**Discussion by Mars-CN** [micropython/micropython#12021] [ESP32] [answered]
In the process of using C to extend MicroPython, the construction method can be specified through make_new, but if the object is being destroyed, what should the destructor (or deinit) be specified through?

Because during the process of constructing objects, malloc needs to be used to create some space, and when the object is destroyed, free release needs to be called

**Comment by jimmo** [comment, marked as answer]
@Mars-CN you need to use `m_new_obj_with_finaliser` to construct the struct instance (so that the GC knows to call your finaliser), and then add a `__del__` entry to your locals dict (this is your destructor/deinit method).

Look at e.g. `extmod/modsocket.c` for an example.

**Reply by Mars-CN** [reply]
I have tried this method, but in fact, MicroPython did not call it__del__ Corresponding function
```C
a = A()
a = A()
```
If written this way, it should be called during the instantiation of the second A object__del__ Method, but not called unless manually called a.__del__ () can only be used

**Reply by jimmo** [reply]
The finaliser isn't called until the garbage collector runs.

**Reply by Mars-CN** [reply]
How can we release resources without any references?

**Reply by Josverl** [reply]
You could use a deinit() method, or define `__enter__` `__exit__` to use with a context manager

To explicitly, or implicitly, perform your first level cleanup.
